### PR TITLE
Block Bindings: Add context needed by block bindings sources during its processing

### PR DESCRIPTION
**Pull request by SantosGuillamot**
Trac ticket: https://core.trac.wordpress.org/ticket/61642

## What?
Block bindings sources sometimes need to access some context that is not initially available in the block. Right now, we are extending it when the source is registered. This pull request moves the logic so it is added when the block bindings are processed. This means:
* Before: Extend context during block bindings source registration. This implies that the blocks extend the context even when they don't use bindings.
* After: Extend context during block bindings processing. This implies that the context is extended ONLY for the blocks where bindings are defined.

## How?
Instead of using the `get_block_type_uses_context` created [back then](https://github.com/WordPress/wordpress-develop/pull/6079) for this use case, use the `available_context` in the block class during bindings processing.

I assume it is not possible to remove that filter because of backward-compatibility.

### Testing instructions

0. Register a custom field in your site to test it. You can use a code snippet like this:
```
register_meta(
	'post',
	'url_custom_field',
	array(
		'show_in_rest' => true,
		'single'       => true,
		'type'         => 'string',
		'default'	   => 'https://wpmovies.dev/wp-content/uploads/2023/04/goncharov-poster-original-1-682x1024.jpeg',
	)
);
```

1. Go to a page and insert an image block.
2. Go to the Code editor and connect the image to the custom field by adding the metadata bindings property. It should look something like this:
```
<!-- wp:image {"id":134,"sizeSlug":"large","linkDestination":"none","metadata":{"bindings":{"url":{"source":"core/post-meta","args":{"key":"url_custom_field"}},"alt":{"source":"core/post-meta","args":{"key":"text_custom_field"}}}}} -->
<figure class="wp-block-image size-large"><img src="https://wpmovies.dev/wp-content/uploads/2023/04/goncharov-poster-original-1-682x1024.jpeg" alt="Alt content" class="wp-image-134"/></figure>
<!-- /wp:image -->
```
3. Save the page and check that in the frontend, the URL defined in the custom field is used.
4. Repeat the process for paragraph, heading, and button blocks.

### CODE REVIEW

File: src/wp-includes/class-wp-block.php [hunk 1]
```diff
@@ -265,6 +265,15 @@ private function process_block_bindings() {
 				continue;
 			}
 
+			// Add the necessary context defined by the source.
+			if ( ! empty( $block_binding_source->uses_context ) ) {
```
Comment (gziolo) [inline src/wp-includes/class-wp-block.php:305]: It didn't occur to me that we could inject the context just as we are about to compute it. Great idea!

I think the fact that we have a formal block binding source instance helps a ton here.

File: src/wp-includes/class-wp-block-bindings-registry.php [hunk 2]
```diff
@@ -189,20 +189,6 @@ public function register( string $source_name, array $source_properties ) {
 
 		$this->sources[ $source_name ] = $source;
 
-		// Adds `uses_context` defined by block bindings sources.
```
Comment (gziolo) [inline src/wp-includes/class-wp-block-bindings-registry.php:192]: There is a single test failure after removing this code:

Screenshot 2024-04-30 at 15 09 16

It's rather expected, so we only need to adjust the test as the block type doesn't change anymore.

**Review by gziolo — COMMENTED** [review]
This looks very promising.

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 2 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-registry.php:192]: You're right. I am not sure if that test makes sense anymore. Maybe we can substitute it for a block that consumes merged context from different sources. Something like this but extended: [link](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/block-bindings/render.php#L124-L158).

If I am not mistaken, this specific test was added to ensure the merge of multiple context done [here](https://github.com/WordPress/wordpress-develop/pull/6456/files#diff-3fd1b8d9e55d4fe09b5f48002d360d3ba9d338d6f33ef3959828424a590b4947L200) was correct. But I think we don't have that problem anymore.

Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-registry.php:192]: I've pushed this commit to illustrate what I mean: [link](https://github.com/WordPress/wordpress-develop/pull/6456/commits/8035804840e6dabe8a102707317491167c65f466).

File: tests/phpunit/tests/block-bindings/render.php [hunk 3]
```diff
@@ -157,6 +157,94 @@ public function test_passing_uses_context_to_source() {
 		);
 	}
 
+	/**
+	 * Tests passing `uses_context` as argument to the source.
+	 *
+	 * @ticket 99999
```
Comment (gziolo) [inline tests/phpunit/tests/block-bindings/render.php]: This needs to change before committing.

File: tests/phpunit/tests/block-bindings/render.php [hunk 4]
```diff
@@ -157,6 +157,94 @@ public function test_passing_uses_context_to_source() {
...
+		);
+		$first_block_render  = $first_block->render();
+		$second_block        = new WP_Block(
+			$second_parsed_block,
+			array(
+				'commonContext'    => 'common context value',
+				'sourceTwoContext' => 'source two context value',
+			)
+		);
+		$second_block_render = $second_block->render();
```
Comment (gziolo) [inline tests/phpunit/tests/block-bindings/render.php]: It's enough to test a single block with the current implementation.

**Review by gziolo — COMMENTED** [review]
We still need to provide an alternative way to expose the `uses_context` defined for the block binding sources to the client. Previously it was handled through the block type definitions exposed on the server. In the revised approach, that information is only available on the server, so that might have some implications on how existing block binding sources work in the editor.

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 2 above)
Comment (gziolo) [reply src/wp-includes/class-wp-block-bindings-registry.php:192]: I left a separate comment. In short, I think it's the right direction.

**Comment by SantosGuillamot** [comment]
I've been looking at exposing the `uses_context` to the client, and I pushed some changes to expose it through editor settings: [link](https://github.com/WordPress/wordpress-develop/pull/6456/commits/18c83aa73dd9356cf9bff9649f99ee2d87f68faf). That way it could be consumed by the client with `getEditorSettings` as explained [here](https://github.com/WordPress/gutenberg/pull/60826#issuecomment-2109964728).

Is it okay to use editor settings for that? Any other alternative?

File: src/wp-includes/block-bindings.php [hunk 5]
```diff
@@ -129,3 +129,6 @@ function get_all_registered_block_bindings_sources() {
 function get_block_bindings_source( string $source_name ) {
 	return WP_Block_Bindings_Registry::get_instance()->get_registered( $source_name );
 }
+
+// Add the `uses_context` to the block editor settings to ensure it can be consumed in the client.
+add_filter( 'block_editor_settings_all', array( 'WP_Block_Bindings_Registry', 'add_uses_context_to_editor_settings' ), 10 );
```
Comment (gziolo) [inline src/wp-includes/block-bindings.php]: This should be moved to one of the existing functions that applies the filter. Adding filters is for plugins and themes when they don't have access to internals of WordPress.

File: src/wp-includes/class-wp-block-bindings-registry.php [hunk 6]
```diff
@@ -305,4 +291,24 @@ public static function get_instance() {
...
+	 * @since 6.6.0
+	 *
+	 * @param array $settings The block editor settings from the `block_editor_settings_all` filter.
+	 * @return array The editor settings with extended block bindings `uses_context`.
+	 */
+	public static function add_uses_context_to_editor_settings( $settings ) {
+		$registered_block_bindings_sources = get_all_registered_block_bindings_sources();
+		foreach ( $registered_block_bindings_sources as $source ) {
+			if ( ! empty( $source->uses_context ) ) {
+				$settings['__experimentalBlockBindings'][ $source->name ]['usesContext'] = $source->uses_context;
```
Comment (gziolo) [inline src/wp-includes/class-wp-block-bindings-registry.php]: In PHP you need to initialize all the arrays at any level.

Should we also pass the `label` to the client?

**Review by gziolo — COMMENTED** [review]
I left some comments related to the idea to expose setting for block bindings together with the editor settings.

> Is it okay to use editor settings for that? Any other alternative?

We are using the same approach for many editor features so that should be acceptable approach.

File: tests/phpunit/tests/block-bindings/render.php (hunk 4 above)
Comment (SantosGuillamot) [reply tests/phpunit/tests/block-bindings/render.php]: You're right that it feels repetitive. I've removed the "multiple sources" test and just rely on the one testing uses_context: [link](https://github.com/SantosGuillamot/wordpress-develop/blob/6f0a931adbfdd7691c1963215a0b32fddb28957a/tests/phpunit/tests/block-bindings/render.php#L124-L158).

 I changed it for another test to ensure that the second source can't access the context defined by the first one. I check and it fails with the previous implementation.

Let me know what you think.

File: src/wp-includes/block-bindings.php (hunk 5 above)
Comment (SantosGuillamot) [reply src/wp-includes/block-bindings.php]: I moved it to the place where (I believe) the editor settings are initialized: [link](https://github.com/WordPress/wordpress-develop/pull/6456/files#diff-edfc853e670c7bfc5a838dc74f75ba5d8b95cb4adbddf0644d9227bcb692e344R651-R667).

File: src/wp-includes/class-wp-block-bindings-registry.php (hunk 6 above)
Comment (SantosGuillamot) [reply src/wp-includes/class-wp-block-bindings-registry.php]: I changed the way arrays are initialized and I added the `label` as well [here](https://github.com/WordPress/wordpress-develop/pull/6456/files#diff-edfc853e670c7bfc5a838dc74f75ba5d8b95cb4adbddf0644d9227bcb692e344R658-R660).

**Comment by SantosGuillamot** [comment]
I've moved the logic to expose bindings in the editor settings to this other pull request: https://github.com/WordPress/wordpress-develop/pull/7020. The idea is to keep this one to change the way the context is handled.

File: tests/phpunit/tests/block-bindings/render.php (hunk 3 above)
Comment (SantosGuillamot) [reply tests/phpunit/tests/block-bindings/render.php]: Added in this commit: https://github.com/WordPress/wordpress-develop/pull/6456/commits/986ff291501e3598efb32e93b50fef42f798eebd

**Review by artemiomorales — APPROVED** [review]
Looks good to me 👍

**Comment by gziolo** [comment]
@peterwilsoncc opened https://github.com/WordPress/wordpress-develop/pull/7226 where he started the process of syncing WordPress packages to WordPress core for the WP 6.7 release. Once that lands, this PR needs to be rebased as it already reports some merge conflicts and it will be ready to commit.

**Comment by cbravobernal** [comment]
> @peterwilsoncc opened #7226 where he started the process of syncing WordPress packages to WordPress core for the WP 6.7 release. Once that lands, this PR needs to be rebased as it already reports some merge conflicts and it will be ready to commit.

Happy to rebase, review and test it once is ready.

**Comment by gziolo** [comment]
It’s time to rebase this branch and commit changes 🚀

**Comment by SantosGuillamot** [comment]
Rebased done 🙂 Let's see if tests pass, but I believe this should be ready. I'm testing now if it works as expected.

**Comment by SantosGuillamot** [comment]
I've been manually testing it and it looks good. If I connect an image to a custom field it shows the connected value in both the editor and in the frontend (server processing).

**Comment by gziolo** [comment]
Committed with https://core.trac.wordpress.org/changeset/59080.

**Pull request closed without merging**